test/dotnet-publish.Tests/PublishDesktopTests.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System.Diagnostics;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.TestFramework;
+using Microsoft.DotNet.Tools.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.DotNet.Tools.Publish.Tests
+{
+    public class PublishDesktopTests : TestBase
+    {
+        [WindowsOnlyTheory]
+        [InlineData("KestrelDesktopWithRuntimes", "http://localhost:20201", null, "libuv.dll", true)]
+        [InlineData("KestrelDesktopWithRuntimes", "http://localhost:20202", "win7-x64", "libuv.dll", true)]
+        [InlineData("KestrelDesktop", "http://localhost:20204", null, "libuv.dll", true)]
+        [InlineData("KestrelDesktop", "http://localhost:20205", "win7-x64", "libuv.dll", true)]
+        public async Task DesktopApp_WithKestrel_WorksWhenPublished(string project, string url, string runtime, string libuvName, bool runnable)
+        {
+            var testInstance = GetTestInstance()
+                .WithLockFiles();
+
+            var publishCommand = new PublishCommand(Path.Combine(testInstance.TestRoot, project), runtime: runtime);
+            var result = await publishCommand.ExecuteAsync();
+
+            result.Should().Pass();
+
+            // Test the output
+            var outputDir = publishCommand.GetOutputDirectory(portable: false);
+            outputDir.Should().HaveFile(libuvName);
+            outputDir.Should().HaveFile(publishCommand.GetOutputExecutable());
+
+            Task exec = null;
+            if (runnable)
+            {
+                var command = new TestCommand(Path.Combine(outputDir.FullName, publishCommand.GetOutputExecutable()));
+                try
+                {
+                    exec = command.ExecuteAsync(url);
+                    NetworkHelper.IsServerUp(url).Should().BeTrue($"Unable to connect to kestrel server - {project} @ {url}");
+                    NetworkHelper.TestGetRequest(url, url);
+                }
+                finally
+                {
+                    command.KillTree();
+                }
+                if (exec != null)
+                {
+                    await exec;
+                }
+            }
+        }
+
+        [WindowsOnlyTheory]
+        [InlineData("KestrelDesktop", "http://localhost:20207")]
```

Comment:
The purpose of the test (in my mind) is to isolate what may be CLI failures by controlling the environment inside the repo. I see it primarily testing publish layout in this new scenario, but could test runnability as well. 

I see it not as being about which scenarios are covered so much as a "cleaner" test approach. For example, if Kestrel ever introduced a packaging bug (undoubtedly would be rare), this test may fail, and we might be wondering what's going on with the CLI. Having  a test that's controlled entirely inside the repo (on top of the kestrel test) would let us quickly exclude a number of possibilities and work toward a solution faster.